sisl/physics/brillouinzone.py

```diff
@@ -110,6 +120,223 @@
+
+        return func
+
+
+@set_module("sisl.physics")
+class ListApply(IteratorApply):
+    __str__ = _apply_str("return list")
+
+    def dispatch(self, method):
+        """ Dispatch the method by returning list of values """
+        iter_func = super().dispatch(method, eta_key="list")
+        @wraps(method)
+        def func(*args, **kwargs):
+            return [v for v in iter_func(*args, **kwargs)]
+        return func
+
+
+@set_module("sisl.physics")
+class OpListApply(IteratorApply):
+    __str__ = _apply_str("return oplist")
+
+    def dispatch(self, method):
+        """ Dispatch the method by returning oplist of values """
+        iter_func = super().dispatch(method, eta_key="oplist")
+        @wraps(method)
+        def func(*args, **kwargs):
+            return oplist(v for v in iter_func(*args, **kwargs))
+        return func
+
+
+@set_module("sisl.physics")
+class ArrayApply(BrillouinZoneParentApply):
+    __str__ = _apply_str("return numpy.ndarray")
+
+    def dispatch(self, method, eta_key="array"):
+        """ Dispatch the method by one array """
+        @wraps(method)
+        def func(*args, wrap=None, eta=False, **kwargs):
+            bz, parent, wrap, eta = self._parse_kwargs(wrap, eta, eta_key=eta_key)
+            k = bz.k
+            w = bz.weight
+
+            # Get first values
+            v = wrap(method(*args, k=k[0], **kwargs), parent=parent, k=k[0], weight=w[0])
+            eta.update()
+
+            # Create full array
+            if v.ndim == 0:
+                a = np.empty([len(k)], dtype=v.dtype)
+            else:
+                a = np.empty((len(k), ) + v.shape, dtype=v.dtype)
+            a[0] = v
+            del v
+
+            for i in range(1, len(k)):
+                a[i] = wrap(method(*args, k=k[i], **kwargs), parent=parent, k=k[i], weight=w[i])
+                eta.update()
+            eta.close()
+
+            return a
+        return func
+
+
+@set_module("sisl.physics")
+class AverageApply(BrillouinZoneParentApply):
+    __str__ = _apply_str("return average")
+
+    def dispatch(self, method):
+        """ Dispatch the method by averaging """
+        @wraps(method)
+        def func(*args, wrap=None, eta=False, **kwargs):
+            bz, parent, wrap, eta = self._parse_kwargs(wrap, eta, eta_key="average")
+            # Do actual average
+            k = bz.k
+            w = bz.weight
+            v = _asoplist(wrap(method(*args, k=k[0], **kwargs), parent=parent, k=k[0], weight=w[0])) * w[0]
+            eta.update()
+            for i in range(1, len(k)):
+                v += _asoplist(wrap(method(*args, k=k[i], **kwargs), parent=parent, k=k[i], weight=w[i])) * w[i]
+                eta.update()
+            eta.close()
+            return v
+
+        return func
+
+
+@set_module("sisl.physics")
+class DataArrayApply(ArrayApply):
+    __str__ = _apply_str("return xarray.DataArray")
+
+    def dispatch(self, method):
+        """ Dispatch the method by summing """
```

Comment:
Update this docstring?

Reply:
thanks!

Reply:
@tfrederiksen what is your opinion on this ;) :wink: